cmd/jujud/util/util.go

```diff
@@ -140,8 +138,47 @@ func ConnectionIsFatal(logger loggo.Logger, conns ...Pinger) func(err error) boo
 	}
 }
 
-// ConnectionIsDead returns true if the given pinger fails to ping.
-var ConnectionIsDead = func(logger loggo.Logger, conn Pinger) bool {
+// ConnectionIsDead returns true if the given Breakable is broken.
+var ConnectionIsDead = func(logger loggo.Logger, conn Breakable) bool {
+	select {
+	case <-conn.Broken():
```

Comment:
Isn't this racy? When we get an error, we want to know if that error was fatal. Before this,
we were sure to be able to tell if the connection is actually still active because we made a Ping
call. By making this change, we are relying on the supposition that if we get an error from
an RPC request and the connection is broken a receive from conn.Broken() immediately
afterwards will succeed.

However I can easily see a scenario where that's not the case - if the act of making a request
triggers the network layer to notice that the connection is down, the api.monitor.run method
may take a while to react and the Broken channel will not be closed by the time we
get around to calling ConnectionIsDead.

I think it would be preferable to have an IsBroken method on Conn (as well as Broken) so
that it has the opportunity to synchronously ask (for example) the rpc.Conn whether it thinks it has broken.